mne_bids/meg_bids.py

```diff
@@ -267,99 +249,114 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
         10`000 points are in the head shape, they are automatically decimated.
+    config : str | None
+        A path to the configuration file to use if the data is from a BTi
+        system.
     overwrite : bool
         If the file already exists, whether to overwrite it.
     verbose : bool
         If verbose is True, this will print a snippet of the sidecar files. If
         False, no content will be printed.
     """
-
-    fname, ext = os.path.splitext(raw_fname)
-    # BTi data is the only file format that does have a file extension
-    if ext == '':
-        ext = '.pdf'
-    ses_path = op.join(output_path, 'sub-%s' % subject_id,
-                       'ses-%s' % session_id)
-    meg_path = op.join(ses_path, 'meg')
-    if not op.exists(output_path):
-        _mkdir_p(output_path)
-    if not op.exists(meg_path):
-        _mkdir_p(meg_path)
-
-    # create the fnames
-    channels_fname = op.join(meg_path, 'sub-%s_ses-%s_task-%s_run-%s'
-                             '_channels.tsv'
-                             % (subject_id, session_id, task, run))
-    events_tsv_fname = op.join(meg_path, 'sub-%s_ses-%s_task-%s_run-%s'
-                               '_events.tsv'
-                               % (subject_id, session_id, task, run))
-    scans_fname = op.join(ses_path,
-                          'sub-%s_ses-%s_scans.tsv' % (subject_id, session_id))
-    fid_fname = op.join(meg_path,
-                        'sub-%s_ses-%s_coordsystem.json'
-                        % (subject_id, session_id))
-    meg_fname = op.join(meg_path,
-                        'sub-%s_ses-%s_meg.json' % (subject_id, session_id))
-    raw_fname_bids = op.join(meg_path, 'sub-%s_ses-%s_task-%s_run-%s_meg%s'
-                             % (subject_id, session_id, task, run, ext))
-
-    orient = orientation[ext]
-    unit = units[ext]
-    manufacturer = manufacturers[ext]
-
-    # KIT systems
-    if ext in ['.con', '.sqd']:
-        raw = io.read_raw_kit(raw_fname, elp=electrode, hsp=hsp,
-                              mrk=hpi, preload=False)
-
-    # Neuromag or converted-to-fif systems
-    elif ext in ['.fif', '.gz']:
-        raw = io.read_raw_fif(raw_fname, preload=False)
-
-    # BTi systems
-    elif ext == '.pdf':
-        if os.path.isfile(raw_fname):
-            raw = io.read_raw_bti(raw_fname, config_fname=config,
-                                  head_shape_fname=hsp,
-                                  preload=False, verbose=verbose)
-
-    # CTF systems
-    elif ext == '.ds':
-        raw = io.read_raw_ctf(raw_fname)
+    if isinstance(raw_file, str):
+        # We must read in the raw data
+        raw = _read_raw(raw_file, electrode=electrode, hsp=hsp, hpi=hpi,
+                        config=config, verbose=verbose)
+        _, ext = _parse_ext(raw_file)
+    elif isinstance(raw_file, BaseRaw):
+        # Only parse the filename for the extension
+        # Assume that if no filename attr exists, it's a fif file.
```

Comment:
why is that? Shouldn't you throw an error?

Reply:
I would avoid magic

Reply:
if you create a Raw object from `RawArray`, there is no associated filename, and since we'll be saving it to `fif` that's why we default to `fif` here